src/utility/copy.cpp

```diff
@@ -208,32 +213,103 @@ CheckRewritten(List *rewritten) {
 	}
 }
 
-bool
+static bool
 CheckPrefix(const char *str, const char *prefix) {
-	while (*prefix) {
-		if (*prefix++ != *str++) {
+	return strncmp(str, prefix, strlen(prefix)) == 0;
+}
+
+static bool
+IsAllowedStatement(CopyStmt *stmt, bool throw_error = false) {
+	int elevel = throw_error ? ERROR : DEBUG4;
+
+	if (stmt->relation) {
+		Relation rel = table_openrv(stmt->relation, AccessShareLock);
+		bool is_duckdb_table = pgduckdb::IsDuckdbTable(rel);
+		bool is_catalog_table = pgduckdb::IsCatalogTable(rel);
+		table_close(rel, NoLock);
+		if (is_catalog_table) {
+			elog(elevel, "DuckDB does not support querying PG catalog tables");
 			return false;
 		}
+
+		if (stmt->is_from && !is_duckdb_table) {
+			elog(elevel, "pg_duckdb does not support COPY ... FROM ... yet for Postgres tables");
+			return false;
+		}
+	}
+
+	if (stmt->filename == NULL) {
+		elog(elevel, "COPY ... TO STDOUT/FROM STDIN is not supported by DuckDB");
+		return false;
 	}
+
 	return true;
 }
-
-const char *
-MakeDuckdbCopyQuery(PlannedStmt *pstmt, const char *query_string, struct QueryEnvironment *query_env) {
-	CopyStmt *copy_stmt = (CopyStmt *)pstmt->utilityStmt;
-	if (!copy_stmt->filename) {
-		return nullptr;
+static bool
+ContainsDuckdbCopyOption(CopyStmt *stmt) {
+	foreach_node(DefElem, defel, stmt->options) {
+		if (strcmp(defel->defname, "format") == 0) {
+			char *fmt = defGetString(defel);
+			if (strcmp(fmt, "parquet") == 0) {
+				return true;
+			} else if (strcmp(fmt, "json") == 0) {
+				return true;
+			}
+		} else if (strcmp(defel->defname, "partition_by") == 0 || strcmp(defel->defname, "use_tmp_file") == 0 ||
+		           strcmp(defel->defname, "overwrite_or_ignore") == 0 || strcmp(defel->defname, "overwrite") == 0 ||
+		           strcmp(defel->defname, "append") == 0 || strcmp(defel->defname, "filename_pattern") == 0 ||
+		           strcmp(defel->defname, "file_extension") == 0 || strcmp(defel->defname, "per_thread_output") == 0 ||
+		           strcmp(defel->defname, "file_size_bytes") == 0 ||
+		           strcmp(defel->defname, "write_partition_columns") == 0 || strcmp(defel->defname, "array") == 0 ||
+		           strcmp(defel->defname, "compression") == 0 || strcmp(defel->defname, "dateformat") == 0 ||
+		           strcmp(defel->defname, "timestampformat") == 0 || strcmp(defel->defname, "nullstr") == 0 ||
+		           strcmp(defel->defname, "prefix") == 0 || strcmp(defel->defname, "suffix") == 0 ||
+		           strcmp(defel->defname, "compression_level") == 0 || strcmp(defel->defname, "field_ids") == 0 ||
+		           strcmp(defel->defname, "row_group_size_bytes") == 0 ||
+		           strcmp(defel->defname, "row_group_size") == 0 ||
+		           strcmp(defel->defname, "row_groups_per_file") == 0) {
```

Comment:
Would be nice to add a `IsOneOf(str[], str)` to make this easier to read :-)
(and the `IsOneOf` could use `IsEq` :-)

Reply:
done